jenkins/check_projects.py

```diff
@@ -39,190 +41,393 @@
+      help='Enable verbose output')
+  return parser.parse_args()
+
+
 def GetConfig(string):
+  """
+  |return| configuration for project settings.
+  """
   if not string:
     return DEFAULT
   elif not os.path.isfile(string):
     raise argparse.ArgumentTypeError('not a file: %s' % string)
   with open(string) as fp:
     return json.loads(fp.read())
 
-
-def CheckProjects(needed, filt, fix=False):
-  projects = LoadProjects(
-      '%s/../jobs' % os.path.dirname(__file__), filt)
-  print >>sys.stderr, 'Checking %d projects for iam bindings:' % len(projects)
-  for role, needs in sorted(needed.items()):
-    print >>sys.stderr, '  %s: %s' % (role, ','.join(Sane(n) for n in needs))
-
-  threads = [
-      threading.Thread(target=Check, args=(p, needed, fix))
-      for p in sorted(projects)
-  ]
-
-  for thread in threads:
-    thread.start()
-
-  for thread in threads:
-    thread.join()
-
-  print >>sys.stderr, '%d already configured, %d updated, %d problems' % (
-      len(completed), len(updated), len(errors))
-
-  fixers = collections.defaultdict(int)
-  unk = ['user:unknown']
-  for project in errors:
-    for name in helpers.get(project,unk):
-      fixers[name] += 1
-    print >>sys.stderr, '  %s: %s' % (
-        project, ','.join(Sane(s) for s in sorted(helpers.get(project, unk))))
-
-  print >>sys.stderr, 'Helpers:'
-  for name, count in sorted(fixers.items(), key=lambda i: i[1]):
-    print >>sys.stderr, '  %s: %s' % (count, Sane(name))
-  sys.exit(1)
-
-
-def Sane(member):
-  if ':' not in member:
-    raise ValueError(member)
-  email = member.split(':')[1]
-  return email.split('@')[0]
-
-
-def LoadProjects(configs, filt):
-  projects = set()
-  regex = re.compile(filt)
-  for dirname, _, files in os.walk(configs):
-    for path in files:
-      full_path = os.path.join(dirname, path)
-      if not regex.match(path):
-        continue
-      with open(full_path) as fp:
-        for project in re.findall('^PROJECT="?([^"\r\n]+)"?$', fp.read(), re.MULTILINE):
-          if '{' not in project:
-            projects.add(project)
-            continue
-          else:
-            raise ValueError(project)
-  return projects
-
-
-def Update(project, role, member):
-  with semaphore:
-    err = subprocess.call(
-        [
-            'gcloud', '-q', 'projects',
-            'add-iam-policy-binding',
-            '--role=%s' % role,
-            '--member=%s' % member,
-            project,
-        ],
-        stdout=null,
-    )
-
-  if not err:
-    with lock:
-      print >>sys.stderr, 'Added %s as %s to %s' % (member, role, project)
-      updated.add(project)
+class Checker(object):
+  def __init__(self, config):
+    self.config = config
+    self.rlExec = RateLimitedExec()
+    self.results = Results()
+
+  def run(self, filt, fix=False):
+    """
+    Checks projects for correct settings.
+    """
+    def check(project, fix):
+      self.results.reportProject(project)
+
+      properties = [
```

Comment:
If you want to keep `ProjectProperty` take a look at `ProjectProperty.__subclasses__` to automatically generate `IAMProperty` and `EnableVmZonalDNS`

http://stackoverflow.com/questions/3862310/how-can-i-find-all-subclasses-of-a-class-given-its-name